### PR TITLE
Add features from m1 and s1

### PR DESCRIPTION
**Pull request by friedger**
Add new interface BitmapValidator

Add implementations of ImageSetter for transitions
Add implementations for ImageViewCallback for colored images based on the source, for day dreams

Add CachedImageFinder and MetadataFactory

### CODE REVIEW

**Comment by blundell** [comment]
this makes sense to me :+1: think @ataulm would appreciate

File: core/src/main/java/com/novoda/pxfetcher/BitmapValidator.java [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package com.novoda.pxfetcher;
+
+import android.graphics.Bitmap;
+import android.view.View;
+
+public interface BitmapValidator {
+    boolean validate(View view, Bitmap bitmap);
```
Comment (ataulm) [inline core/src/main/java/com/novoda/pxfetcher/BitmapValidator.java]: Can you add doc here or rename? `BitmapValidator.bitmapIsEligibleForLoadingInView(View, Bitmap)`?

File: core/src/main/java/com/novoda/pxfetcher/CrossFadableDrawable.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
+package com.novoda.pxfetcher;
+
+import android.graphics.drawable.Drawable;
+import android.graphics.drawable.TransitionDrawable;
+
+public class CrossFadableDrawable extends TransitionDrawable {
```
Comment (ataulm) [inline core/src/main/java/com/novoda/pxfetcher/CrossFadableDrawable.java]: CrossFadeDrawable (it's not that the drawable is fadable - it definitely will fade)

File: core/src/main/java/com/novoda/pxfetcher/CrossFadableDrawable.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        drawable.setId(0, PREV_ID);
+        drawable.setId(1, NEXT_ID);
+        return drawable;
+    }
+
+    CrossFadableDrawable(Drawable[] layers) {
+        super(layers);
+    }
+
+    public Drawable getPrev() {
```
Comment (ataulm) [inline core/src/main/java/com/novoda/pxfetcher/CrossFadableDrawable.java]: `getFirstDrawable/getSecondDrawable()`

`getDrawableFrom()/getDrawableTo()`

`getSourceDrawable()/getDestinationDrawable()`

File: core/src/main/java/com/novoda/pxfetcher/CrossFadableDrawable.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        return findDrawableByLayerId(PREV_ID);
+    }
+
+    public Drawable getNext() {
+        return findDrawableByLayerId(NEXT_ID);
+    }
+
+    public void setPrev(Drawable drawable) {
```
Comment (ataulm) [inline core/src/main/java/com/novoda/pxfetcher/CrossFadableDrawable.java]: Are these setters used in m1/s1? If not, I'd suggesting removing until we need it.

File: core/src/main/java/com/novoda/pxfetcher/DebugImageViewCallback.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    public void onResult(Success ok) {
+        ImageView imageView = imageViewWeakReference.get();
+        if (imageView == null) {
+            return;
+        }
+
+        Resources resources = imageView.getResources();
+        int color;
+        if (ok instanceof MemoryRetriever.Success) {
+            color = resources.getColor(R.color.debug_imageloader_memory_overlay);
```
Comment (ataulm) [inline core/src/main/java/com/novoda/pxfetcher/DebugImageViewCallback.java]: we should prefix these resources with `px__` to allow the users to override (intentionally)

File: core/src/main/java/com/novoda/pxfetcher/FadingImageTransition.java [hunk 6]
```diff
@@ -0,0 +1,27 @@
+package com.novoda.pxfetcher;
+
+import android.graphics.drawable.Drawable;
+import android.graphics.drawable.TransitionDrawable;
+import android.widget.ImageView;
+
+class FadingImageTransition implements ImageTransition {
```
Comment (ataulm) [inline core/src/main/java/com/novoda/pxfetcher/FadingImageTransition.java]: `CrossFadeImageTransition`?

File: core/src/main/java/com/novoda/pxfetcher/FadingImageSetter.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
+package com.novoda.pxfetcher;
+
+import android.content.res.Resources;
+import android.graphics.Bitmap;
+import android.graphics.drawable.BitmapDrawable;
+import android.graphics.drawable.Drawable;
+import android.widget.ImageView;
+
+public class FadingImageSetter implements ImageSetter {
```
Comment (ataulm) [inline core/src/main/java/com/novoda/pxfetcher/FadingImageSetter.java]: `CrossFadeImageSetter`?

File: core/src/main/res/values/colors.xml [hunk 8]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+  <color name="debug_imageloader_memory_overlay">#8040FF40</color>
```
Comment (ataulm) [inline core/src/main/res/values/colors.xml]: ah cool we're defining here - can we rename to add a prefix for imageloader/pxfetcher

**Comment by friedger** [comment]
CrossFadeImageSetter and FadingImageSetter have two differences:
In CrossFadeImageSetter a new Drawable(0000000) is used if the transition can't be run.
In FadingImageSetter the transitionDrawable is crossfadeEnabled.
These have been merged together.

**Comment by ataulm** [comment]
:+1:

**Pull request merged**